[libpas] Use pas_enumerator_copy_remote() wrapper instead of raw memory reader in PGM crash reporting

zakariaridouh opened the pull request:
#### 98df8ea43a4f9b671603e586762aad47d75a707b
<pre>
[libpas] Use pas_enumerator_copy_remote() wrapper instead of raw memory reader in PGM crash reporting
<a href="https://bugs.webkit.org/show_bug.cgi?id=299403">https://bugs.webkit.org/show_bug.cgi?id=299403</a>
<a href="https://rdar.apple.com/161206379">rdar://161206379</a>

Reviewed by Brandon Stewart and Dan Hecht.

Use the safer and more used `pas_enumerator_copy_remote()` wrapper to read memory
instead of rolling our own memory reader.

* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_enumerator_reader_adapter):
(setup_enumerator_for_crash_reporting):
(pas_update_report_crash_fields_and_cleanup):
(pas_report_crash_extract_pgm_failure):
(memory_reader_adapter): Deleted.
(setup_memory_reader): Deleted.

Canonical link: <a href="https://commits.webkit.org/301166@main">https://commits.webkit.org/301166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac297f55eacb2309c2982aa2f27d16ef8dd11266

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125112 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44779 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131960 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1931ed6d-d8f3-483a-899f-eb948da43964) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45470 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53343 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95237 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ddd2a784-cdac-44bb-a0cb-688a9c95dade) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36296 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111888 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75779 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/50a1936e-2436-4c48-95d2-d40faad5a85c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35206 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30051 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75447 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117203 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106061 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30281 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134639 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123630 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51923 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39715 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103708 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52356 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108109 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103479 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26351 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48825 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27122 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48976 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51815 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156653 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51182 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39233 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54538 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52872 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -82,95 +92,88 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
+                pas_enumerator_destroy(enumerator);
                 return KERN_FAILURE;
+            }
+            resolved_dealloc_backtrace = &dealloc_backtrace_data;
```
any reason to not just use `&dealloc_backtrace_data` below and remove `resolved_dealloc_backtrace`?

zakariaridouh replied:
Updated. `resolved_dealloc_backtrace` was leftover from previous way of reading memory, removed it!

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -76,100 +87,91 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
-            if (resolved_dealloc_backtrace->frame_size < 0 || resolved_dealloc_backtrace->frame_size > PGM_BACKTRACE_MAX_FRAMES)
+            }
+            if (dealloc_backtrace_data.frame_size < 0 || dealloc_backtrace_data.frame_size > PGM_BACKTRACE_MAX_FRAMES) {
+                pas_enumerator_destroy(enumerator);
```
I noticed quite a few instances of the below code repeated over again; not sure if we can simplify this at all.

```
pas_enumerator_destroy(enumerator);
return KERN_FAILURE;
```

zakariaridouh replied:
Yeah, there's some repetitive logic. I've been meaning to do a small rewrite/cleanup of the logic and functions as a followup patch, but I don't think this should block landing this patch.

zakariaridouh replied:
Put a patch up for this here: https://github.com/WebKit/WebKit/pull/52339

zakariaridouh committed:
[libpas] Use pas_enumerator_copy_remote() wrapper instead of raw memory reader in PGM crash reporting
https://bugs.webkit.org/show_bug.cgi?id=299403
rdar://161206379

Reviewed by Brandon Stewart and Dan Hecht.

Use the safer and more used `pas_enumerator_copy_remote()` wrapper to read memory
instead of rolling our own memory reader.

* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_enumerator_reader_adapter):
(setup_enumerator_for_crash_reporting):
(pas_update_report_crash_fields_and_cleanup):
(pas_report_crash_extract_pgm_failure):
(memory_reader_adapter): Deleted.
(setup_memory_reader): Deleted.

Canonical link: https://commits.webkit.org/301166@main
